cmake/detect_gpu.cmake

```diff
@@ -0,0 +1,215 @@
+        "  for (int device = 0; device < count; ++device)\n"
+        "  {\n"
+        "    cudaDeviceProp prop;\n"
+        "    if(cudaSuccess == cudaGetDeviceProperties(&prop, device))\n"
+        "      std::cout << prop.major << prop.minor << ';';"
+        "  }\n"
+        "  return 0;\n"
+        "}\n")
+
+    execute_process(COMMAND "${CMAKE_CUDA_COMPILER}" "--run" "${__cufile}"
+                    WORKING_DIRECTORY "${PROJECT_BINARY_DIR}/CMakeFiles/"
+                    RESULT_VARIABLE __nvcc_res OUTPUT_VARIABLE __nvcc_out
+                    ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    if(__nvcc_res EQUAL 0)
+        set(CUDA_gpu_detect_output ${__nvcc_out} CACHE INTERNAL
+            "Returned GPU architetures from ginkgo_detect_gpus tool" FORCE)
+    endif()
+
+    if(NOT CUDA_gpu_detect_output)
+        message(STATUS "Automatic GPU detection failed.
+                       Building for all known architectures.")
+        set(${out_variable} ${ginkgo_known_gpu_archs} PARENT_SCOPE)
+    else()
+        set(${out_variable} ${CUDA_gpu_detect_output} PARENT_SCOPE)
+    endif()
+endfunction()
+
+
+################################################################################
+# A function for checking whether it is available
+# Usage:
+#     check_available(arch_var)
+function(check_available arch_var)
+    if(arch_var MATCHES "([0-9]+)")
+        if(NOT(arch_var IN_LIST cuda_all_arch_list))
+            message(FATAL_ERROR "arch ${arch_var} is not invalid")
+        elseif(arch_var IN_LIST ginkgo_known_gpu_archs)
+            if(arch_var IN_LIST ginkgo_unsupported_archs)
+                message(FATAL_ERROR "Ginkgo does not support ${arch_var}")
+            endif()
+        else()
+            message(FATAL_ERROR
+                "CUDA ${CMAKE_CUDA_COMPILER_VERSION} does not know ${arch_var}")
+        endif()
+    else()
+        if(arch_var IN_LIST ginkgo_known_gpu_archs_name)
+            if(arch_var IN_LIST ginkgo_unsupported_archs_name)
+                message(FATAL_ERROR
+                    "Ginkgo does not support ${arch_var}")
+            endif()
+        else ()
+            message(FATAL_ERROR
+                "CUDA ${CMAKE_CUDA_COMPILER_VERSION} does not know ${arch_var}")
+        endif()
+    endif()
+endfunction()
+
+
+################################################################################
+# Function for selecting GPU arch flags for nvcc based on CUDA_ARCH_OPTION
+# Usage:
+#     ginkgo_select_nvcc_arch_flags(out_variable)
+function(ginkgo_select_nvcc_arch_flags out_variable)
+    if(CMAKE_CUDA_COMPILER_VERSION MATCHES "([0-9]+).([0-9]+).(.*)")
+        set(cuda_version ${CMAKE_MATCH_1}${CMAKE_MATCH_2})
+    endif()
+    set(ginkgo_known_gpu_archs ${ginkgo_known_gpu_archs_cuda_${cuda_version}})
+    set(ginkgo_known_gpu_archs_name
+        ${ginkgo_known_gpu_archs_name_${cuda_version}})
+    set(cuda_arch_bin_Volta ${cuda_arch_bin_Volta_${cuda_version}})
+    set(cuda_all_arch_list "")
```

Comment:
nit: is this variable needed somewhere?

Reply:
I use cuda_all_arch_list to verify whether arch is valid.

Reply:
Ah, I see now, could you the pass it as an argument to `check_variable`?
I know that `cmake` doesn't require you to do it, since its functions are macro-like constructs (somewhat similar to `#define function(...) { /*function body*/ }` in C and C++). But this is very unusual behavior and makes the code hard to follow. That's why I would prefer if we didn't rely on it.